### PR TITLE
auto invite from payment link

### LINKED ISSUE #497: Auto invite from payment links

### LINKED ISSUE #605: Invite page flashes offline
Introduced by splash screen change. Very fugly.

https://github.com/daimo-eth/daimo/assets/6984346/a4dce283-51ea-402e-b30c-2175c9105eba

### PR DESCRIPTION
**Pull request by nalinbhardwaj**
https://streamable.com/gjcast

Enables new short payment links.
Starts refactoring code based invites as well.

Closes #497.
Closes #605.

### CODE REVIEW

**Comment by dcposch** [comment]
## Tell user to paste link

In the demo video, app requests clipboard permission as soon as the user taps the Invite Code textbox. This is a bit surprising, and I think many users won't tap the textbox--will just perceive they're stuck missing an invite code.

I'd suggest a button on one or both of these two pages. Something along the lines of

> **Accept Payment Link**
> Create a new account and claim USDC that someone sent you.

image image

**Comment by nalinbhardwaj** [comment]
> I'd suggest a button on one or both of these two pages. Something along the lines of
> 
> > **Accept Payment Link**
> > Create a new account and claim USDC that someone sent you.

That seems like a very loaded concept to introduce in onboarding. Alternately, treating it as an "invite" and making the payment link get claimed under-the-hood post-onboarding feels more delightful to me.

Some thoughts:
- The web page is intended to emphasise that you're copying an invite to your clipboard, thoughts on improving emphasis there instead?
- Branch links work the same way (they actually prompt you to paste as soon as you first open, even without any fields/tap):
    - 
- We could add a small paste button next to the text bar, thoughts on that?

**Comment by dcposch** [comment]
*Branch links work the same way (they actually prompt you to paste as soon as you first open, even without any fields/tap)*

Yeah, that makes sense--this is different. Prompts the user immediately. If branch opened to a page that said Invite Code [ empty textbox ] users would get confused ("where do i get an invite code?") and churn.

*We could add a small paste button next to the text bar, thoughts on that?*

Sounds great. Eg small text, [Paste invite code from link.](https://www.youtube.com/watch?v=dQw4w9WgXcQ)

**Comment by nalinbhardwaj** [comment]
> _We could add a small paste button next to the text bar, thoughts on that?_
> 
> Sounds great. Eg small text, [Paste invite code from link.](https://www.youtube.com/watch?v=dQw4w9WgXcQ)

Done, new flow video: https://streamable.com/nb6hxg


Three states of the screen: 1) pre-typing 2) pasting a payment link 3) typing in an invite code


Screenshot 2024-01-09 at 10 19 27 PM
Screenshot 2024-01-09 at 10 23 07 PM
Screenshot 2024-01-09 at 10 24 08 PM

File: apps/daimo-mobile/src/sync/sync.ts [hunk 1]
```diff
@@ -58,9 +58,11 @@ function hasPendingOps(account: Account) {
   );
 }
 
-async function maybeSync(fromScratch?: boolean) {
+type SyncStatus = "success" | "failed" | "skipped" | "skipped";
+
+async function maybeSync(fromScratch?: boolean): Promise<SyncStatus> {
   const manager = getAccountManager();
-  if (manager.currentAccount == null) return;
+  if (manager.currentAccount == null) return "skipped";
```
Comment (nalinbhardwaj) [inline apps/daimo-mobile/src/sync/sync.ts:65]: @wojtus7 there was a bug here during onboarding due to the splash screen change: #605

File: apps/daimo-mobile/src/sync/sync.ts [hunk 2]
```diff
@@ -25,11 +25,11 @@ import { Account, getAccountManager } from "../model/account";
...
   maybeSync(true)
-    .then((success) => {
-      if (!success) {
+    .then((status) => {
+      if (status === "failed") {
         updateNetworkState(() => {
-          // set fail to 3 because at 2 it would remove offline banner
-          // for one try even if sync failed
+          // HACK: mark offline if splash screen sync fails due to network state
+          // by setting syncAttemptsFailed to 3
```
Comment (dcposch) [inline apps/daimo-mobile/src/sync/sync.ts:32]: dont think the hack should be necessary--can check, but just

`{ status: "offline", syncAttemptsFailed: 1 }` should do it imo

File: apps/daimo-mobile/src/view/screen/onboarding/InvitePage.tsx [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+  ) => void;
+}) {
+  // Handle invite links from deep link opens.
+  const [text, setText] = useState(
+    inviteLink ? formatDaimoLink(inviteLink) : ""
+  );
+  const [isValid, setIsValid] = useState(!!inviteLink);
+  const [sender, setSender] = useState<EAccount | undefined>(undefined);
+
+  // HACK: Workaround for testnet to never query API
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/onboarding/InvitePage.tsx:49]: long term we can fix this by merging the APIs.
single API > can multiplex on backend

File: apps/daimo-mobile/src/view/screen/onboarding/InvitePage.tsx [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    if (!linkStatus || linkStatus.data == null) return;
+
+    if (linkStatus.data.link.type === "notev2") {
+      const noteStatus = linkStatus.data as DaimoNoteStatus;
+      setIsValid(noteStatus.status === DaimoNoteState.Confirmed);
+      setSender(noteStatus.sender);
+    } else if (linkStatus.data.link.type === "invite") {
+      const inviteStatus = linkStatus.data as DaimoInviteStatus;
+      setIsValid(inviteStatus.isValid);
+      setSender(undefined); // TODO: Add senders to invite codes
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/onboarding/InvitePage.tsx:75]: 👍

File: apps/daimo-mobile/src/view/screen/onboarding/InvitePage.tsx [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        <Spacer h={16} />
+        {hasNotStartedTyping ? (
+          <>
+            <TextCenter>
+              <TextLight>or</TextLight>
+            </TextCenter>
+            <Spacer h={16} />
+            <ButtonBig
+              type="primary"
+              title="Paste invite from link"
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/onboarding/InvitePage.tsx:154]: 👍

File: apps/daimo-web/src/app/link/[[...slug]]/CallToAction.tsx [hunk 6]
```diff
@@ -45,7 +48,8 @@ export function CallToAction({
           href={directDeepLink}
           className="block text-center text-primaryLight tracking-wider font-bold py-5"
         >
-          ALREADY HAVE IT? OPEN IN DAIMO
+          ALREADY HAVE IT? OPEN {isInvitePaymentLink ? "INVITE" : "LINK"} IN
+          DAIMO
```
Comment (dcposch) [inline apps/daimo-web/src/app/link/[[...slug]]/CallToAction.tsx:52]: 👍

File: apps/daimo-web/src/components/buttons.tsx [hunk 7]
```diff
@@ -1,22 +1,45 @@
...
+        (justCopied ? "bg-success" : "bg-primaryLight") +
+        " tracking-wider text-white font-bold py-5 w-full rounded-md disabled:opacity-50"
+      }
       disabled={disabled}
       onClick={onClick}
     >
-      GET DAIMO
+      {justCopied
+        ? "COPIED, REDIRECTING..."
+        : (inviteDeepLink ? "COPY INVITE AND " : "") + "GET DAIMO"}
```
Comment (dcposch) [inline apps/daimo-web/src/components/buttons.tsx:42]: 👍

File: apps/daimo-mobile/src/view/screen/onboarding/InvitePage.tsx (hunk 4 above)
Comment (dcposch) [reply apps/daimo-mobile/src/view/screen/onboarding/InvitePage.tsx:75]: looks like inviteStatus already has `inviteStatus.sender`

**Review by dcposch — APPROVED** [review]
LGTM

File: apps/daimo-mobile/src/view/screen/onboarding/InvitePage.tsx (hunk 4 above)
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/view/screen/onboarding/InvitePage.tsx:75]: yeah not populated currently / need to add a proper backend for invites and their graph

File: apps/daimo-mobile/src/action/useCreateAccount.ts [hunk 8]
```diff
@@ -73,6 +89,29 @@ export function useCreateAccount(
...
+        );
+        const claimTxHash = await rpcFunc.claimEphemeralNoteSponsored.mutate({
+          ephemeralOwner: noteStatus.ephemeralOwner!,
+          recipient: address,
+          signature: ephemeralSignature,
+        });
+
+        console.log(`[CREATE] Claimed note in background: ${claimTxHash}`);
+      };
+      claimNote();
```
[the remark this answers is not in the document]
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/action/useCreateAccount.ts]: good catch, added

File: apps/daimo-mobile/src/logic/invite.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
+import {
+  DaimoLinkInvite,
+  DaimoLinkNoteV2,
+  parseDaimoLink,
+} from "@daimo/common";
+
+/* Interpret text as a potential invite link */
```
[the remark this answers is not in the document]
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/logic/invite.ts:7]: done, I was mostly pre-empting in the current code that we'll enable this pattern for all types of links in future

File: apps/daimo-mobile/src/sync/sync.ts (hunk 2 above)
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/sync/sync.ts]: done, this was old code -- I think it was trying to make the first load show offline by itself, but we don't need that, fair.

File: apps/daimo-mobile/src/view/screen/onboarding/InvitePage.tsx (hunk 3 above)
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/view/screen/onboarding/InvitePage.tsx:50]: pulled into its own issue

**Pull request merged**